src/mono/mono/mini/mini-s390x.c

```diff
@@ -4834,605 +5502,516 @@ mono_arch_output_basic_block (MonoCompile *cfg, MonoBasicBlock *bb)
+			s390_vlgvf (code, s390_r13, ins->sreg2, 0, 0);
+			s390_vlvgf (code, ins->dreg, s390_r13, 0, GTMREG_TO_UINT32 (ins->inst_c0));
 			break;
-		case OP_PSHRQ_REG:
-			s390x_psrlq (code, ins->dreg, ins->sreg2);
+		case OP_INSERT_R8:
+			s390_vlgvg (code, s390_r13, ins->sreg2, 0, 0);
+			s390_vlvgg (code, ins->dreg, s390_r13, 0, GTMREG_TO_UINT32 (ins->inst_c0));
 			break;
-
-		/*TODO: This is appart of the sse spec but not added
-		case OP_PSARQ:
-			s390x_psraq_reg_imm (code, ins->dreg, ins->inst_imm);
+		case OP_EXTRACT_I1:
+			s390_vlgvb (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
 			break;
-		case OP_PSARQ_REG:
-			s390x_psraq (code, ins->dreg, ins->sreg2);
+		case OP_EXTRACT_I2:
+			s390_vlgvh (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
 			break;
-		*/
-
-		case OP_PSHLQ:
-			s390x_psllq_reg_imm (code, ins->dreg, ins->inst_imm);
+		case OP_EXTRACT_I4:
+			s390_vlgvf (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
 			break;
-		case OP_PSHLQ_REG:
-			s390x_psllq (code, ins->dreg, ins->sreg2);
+		case OP_EXTRACT_I8:
+			s390_vlgvg (code, ins->dreg, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
 			break;
-		case OP_CVTDQ2PD:
-			s390x_cvtdq2pd (code, ins->dreg, ins->sreg1);
+		case OP_EXTRACT_R4:
+			s390_vlgvf (code, s390_r13, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+			s390_vlvgf (code, ins->dreg, s390_r13, 0, 0);
 			break;
-		case OP_CVTDQ2PS:
-			s390x_cvtdq2ps (code, ins->dreg, ins->sreg1);
+		case OP_EXTRACT_R8:
+			s390_vlgvg (code, s390_r13, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+			s390_ldgr (code, ins->dreg, s390_r13);
 			break;
-		case OP_CVTPD2DQ:
-			s390x_cvtpd2dq (code, ins->dreg, ins->sreg1);
+		case OP_XEXTRACT_I1:
+			s390_vlgvb (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
 			break;
-		case OP_CVTPD2PS:
-			s390x_cvtpd2ps (code, ins->dreg, ins->sreg1);
+		case OP_XEXTRACT_I2:
+			s390_vlgvh (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
 			break;
-		case OP_CVTPS2DQ:
-			s390x_cvtps2dq (code, ins->dreg, ins->sreg1);
+		case OP_XEXTRACT_I4:
+			s390_vlgvf (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
 			break;
-		case OP_CVTPS2PD:
-			s390x_cvtps2pd (code, ins->dreg, ins->sreg1);
+		case OP_XEXTRACT_I8:
+			s390_vlgvg (code, ins->dreg, ins->sreg1, ins->sreg2, 0);
 			break;
-		case OP_CVTTPD2DQ:
-			s390x_cvttpd2dq (code, ins->dreg, ins->sreg1);
+		case OP_XEXTRACT_R4:
+			s390_vlgvf (code, s390_r13, ins->sreg1, ins->sreg2, 0);
+			s390_ldgr (code, ins->dreg, s390_r13);
 			break;
-		case OP_CVTTPS2DQ:
-			s390x_cvttps2dq (code, ins->dreg, ins->sreg1);
+		case OP_XEXTRACT_R8:
+			s390_vlgvg (code, s390_r13, ins->sreg1, ins->sreg2, 0);
+			s390_ldgr (code, ins->dreg, s390_r13);
 			break;
-
-		case OP_ICONV_TO_X:
-			amd64_movd_xreg_reg_size (code, ins->dreg, ins->sreg1, 4);
+		case OP_EXPAND_I1:
+			s390_vlvgb (code, s390_vr16, ins->sreg1, 0, GTMREG_TO_UINT32 (ins->inst_c0));
+			s390_vrepb (code, ins->dreg, s390_vr16, 0);
```

Comment:
Why do we need vr16 here?  Can't this just load into `ins->dreg` and then replicate in-place?